pkg/auth/dependency/welcomemessage/provider.go

```diff
@@ -0,0 +1,83 @@
+package welcomemessage
+
+import (
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/identity"
+	taskspec "github.com/skygeario/skygear-server/pkg/auth/task/spec"
+	"github.com/skygeario/skygear-server/pkg/core/async"
+	"github.com/skygeario/skygear-server/pkg/core/auth/metadata"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/mail"
+	"github.com/skygeario/skygear-server/pkg/core/template"
+)
+
+type Provider struct {
+	AppName                     string
+	EmailConfig                 config.EmailMessageConfiguration
+	WelcomeMessageConfiguration *config.WelcomeMessageConfiguration
+	TemplateEngine              *template.Engine
+	TaskQueue                   async.Queue
+}
+
+func (p *Provider) send(emails []string) (err error) {
+	if !p.WelcomeMessageConfiguration.Enabled {
+		return
+	}
+
+	if p.WelcomeMessageConfiguration.Destination == config.WelcomeMessageDestinationFirst {
+		emails = emails[0:1]
```

Comment:
The `emails` list here maybe empty, if the identities don't have email claims?

Reply:
The slice behavior is unlike Python :(